### PR TITLE
feat(CLI): move soroban-spec-json/typescipt; add soroban-spec-tools

### PR DESCRIPTION
**Pull request by willemneal**
### What

Move the crates `soroban-spec-json` and `soroban-spec-typescript` to this repo and move `strval.rs`  a new crate called sorobon-spec-tools.

### Why

Remove dependency on sdk and allow for using the JSON parsing outside the CLI.

### Known limitations

[TODO or N/A]

### CODE REVIEW

**Review by leighmcculloch — COMMENTED** [review]
Reviewing just the soroban-spec-json copy from rs-soroban-sdk, I noticed it looks like something happened and it isn't just a copy-paste. A dependency on stellar-xdr has changed to soroban-env-host, which is unexpected. soroban-spec-json shouldn't depend on the entire host, it only needs xdr.

I'll take another look tomorrow. It would be ideal if for copying the crate it was mostly just a copy-paste without mods.

I defer to @paulbellamy and co for the remainder.

**Comment by willemneal** [comment]
My thought here was that I wanted to ensure that all crates used the same xdr. That is if we use stellar-xdr directly they could get out of sync.

**Comment by leighmcculloch** [comment]
@willemneal We use cargo-deny on our other Rust repos to make sure we aren't depending on multiple versions of any crate, not just stellar-xdr. We could add it here to address that concern.

**Comment by willemneal** [comment]
@leighmcculloch I'm getting errors when I tried to switch over. A very easy solution would be to allow soroban_spec to re-export the xdr.

I'm currently getting errors like the following and hope to repro on CI:

```
error[E0432]: unresolved import `stellar_xdr::StringM`
   --> /Users/willem/.cargo/git/checkouts/rs-soroban-env-230c62c6b1f3d3f1/0c0cae6/soroban-env-common/src/symbol.rs:138:5
    |
138 | use stellar_xdr::StringM;
    |     ^^^^^^^^^^^^^^^^^^^^ no `StringM` in the root
    |
help: consider importing one of these items instead
    |
138 | use crate::xdr::curr::StringM;
    |     ~~~~~~~~~~~~~~~~~~~~~~~~~
138 | use crate::xdr::next::StringM;
    |     ~~~~~~~~~~~~~~~~~~~~~~~~~
138 | use stellar_xdr::curr::StringM;
    |     ~~~~~~~~~~~~~~~~~~~~~~~~~~
138 | use stellar_xdr::next::StringM;
    |     ~~~~~~~~~~~~~~~~~~~~~~~~~~

error[E0432]: unresolved import `stellar_xdr::SCSYMBOL_LIMIT`
   --> /Users/willem/.cargo/git/checkouts/rs-soroban-env-230c62c6b1f3d3f1/0c0cae6/soroban-env-common/src/symbol.rs:139:5
    |
139 | use stellar_xdr::SCSYMBOL_LIMIT;
    |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^ no `SCSYMBOL_LIMIT` in the root

error[E0432]: unresolved imports `stellar_xdr::Int256Parts`, `stellar_xdr::ScVal`, `stellar_xdr::UInt256Parts`
  --> /Users/willem/.cargo/git/checkouts/rs-soroban-env-230c62c6b1f3d3f1/0c0cae6/soroban-env-common/src/num.rs:10:19
   |
10 | use stellar_xdr::{Int256Parts, ScVal, UInt256Parts};
   |                   ^^^^^^^^^^^  ^^^^^  ^^^^^^^^^^^^ no `UInt256Parts` in the root
   |                   |            |
   |                   |            no `ScVal` in the root
   |                   no `Int256Parts` in the root
   |
   = help: consider importing one of these items instead:
           crate::xdr::next::Int256Parts
           crate::xdr::next::Type::Int256Parts
           crate::xdr::next::TypeVariant::Int256Parts
           stellar_xdr::next::Int256Parts
           stellar_xdr::next::Type::Int256Parts
           stellar_xdr::next::TypeVariant::Int256Parts
   = help: consider importing one of these items instead:
           crate::xdr::next::ScVal
           crate::xdr::next::Type::ScVal
           crate::xdr::next::TypeVariant::ScVal
           stellar_xdr::next::ScVal
           stellar_xdr::next::Type::ScVal
           stellar_xdr::next::TypeVariant::ScVal
   = help: consider importing one of these items instead:
           crate::xdr::next::Type::UInt256Parts
           crate::xdr::next::TypeVariant::UInt256Parts
           crate::xdr::next::UInt256Parts
           stellar_xdr::next::Type::UInt256Parts
           stellar_xdr::next::TypeVariant::UInt256Parts
           stellar_xdr::next::UInt256Parts

error[E0432]: unresolved imports `crate::xdr::ScSymbol`, `crate::xdr::ScVal`
   --> /Users/willem/.cargo/git/checkouts/rs-soroban-env-230c62c6b1f3d3f1/0c0cae6/soroban-env-common/src/symbol.rs:378:18
    |
378 | use crate::xdr::{ScSymbol, ScVal};
    |                  ^^^^^^^^  ^^^^^ no `ScVal` in the root
    |                  |
    |                  no `ScSymbol` in the root
    |
    = help: consider importing one of these items instead:
            crate::xdr::next::ScSymbol
            crate::xdr::next::Type::ScSymbol
            crate::xdr::next::TypeVariant::ScSymbol
            stellar_xdr::next::ScSymbol
            stellar_xdr::next::Type::ScSymbol
            stellar_xdr::next::TypeVariant::ScSymbol
    = help: consider importing one of these items instead:
            crate::xdr::next::ScVal
            crate::xdr::next::Type::ScVal
            crate::xdr::next::TypeVariant::ScVal
            stellar_xdr::next::ScVal
            stellar_xdr::next::Type::ScVal
            stellar_xdr::next::TypeVariant::ScVal

error[E0432]: unresolved import `crate::xdr::ScHostContextErrorCode`
 --> /Users/willem/.cargo/git/checkouts/rs-soroban-env-230c62c6b1f3d3f1/0c0cae6/soroban-env-common/src/vmcaller_env.rs:2:5
  |
2 | use crate::xdr::ScHostContextErrorCode;
  |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ no `ScHostContextErrorCode` in the root
  |
help: consider importing one of these items instead
  |
2 | use crate::xdr::next::ScHostContextErrorCode;
  |     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
2 | use crate::xdr::next::Type::ScHostContextErrorCode;
  |     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
2 | use crate::xdr::next::TypeVariant::ScHostContextErrorCode;
  |     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
2 | use stellar_xdr::next::ScHostContextErrorCode;
  |     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    and 2 other candidates

error[E0433]: failed to resolve: could not find `Error` in `stellar_xdr`
   --> /Users/willem/.cargo/git/checkouts/rs-soroban-env-230c62c6b1f3d3f1/0c0cae6/soroban-env-common/src/status.rs:187:41
    |
187 |                 return Err(stellar_xdr::Error::Invalid);
    |                                         ^^^^^ could not find `Error` in `stellar_xdr`

error[E0425]: cannot find function, tuple struct or tuple variant `ScSymbol` in crate `crate::xdr`
   --> /Users/willem/.cargo/git/checkouts/rs-soroban-env-230c62c6b1f3d3f1/0c0cae6/soroban-env-common/src/env_val.rs:338:46
    |
338 |                 Ok(ScVal::Symbol(crate::xdr::ScSymbol(
    |                                              ^^^^^^^^ not found in `crate::xdr`
    |
help: consider importing one of these items
    |
1   + use crate::xdr::next::ScSymbol;
    |
1   + use crate::xdr::next::Type::ScSymbol;
    |
1   + use crate::xdr::next::TypeVariant::ScSymbol;
    |
1   + use stellar_xdr::next::ScSymbol;
    |
      and 2 other candidates
help: if you import `ScSymbol`, refer to it directly
    |
338 -                 Ok(ScVal::Symbol(crate::xdr::ScSymbol(
338 +                 Ok(ScVal::Symbol(ScSymbol(
    |

error[E0412]: cannot find type `Error` in crate `stellar_xdr`
   --> /Users/willem/.cargo/git/checkouts/rs-soroban-env-230c62c6b1f3d3f1/0c0cae6/soroban-env-common/src/raw_val.rs:341:24
    |
341 | impl From<stellar_xdr::Error> for ConversionError {
    |                        ^^^^^ not found in `stellar_xdr`
    |
help: consider importing one of these items
    |
1   + use core::error::Error;
    |
1   + use core::fmt::Error;
    |
1   + use crate::xdr::curr::Error;
    |
1   + use crate::xdr::next::Error;
    |
      and 11 other candidates
help: if you import `Error`, refer to it directly
    |
341 - impl From<stellar_xdr::Error> for ConversionError {
341 + impl From<Error> for ConversionError {
    |

error[E0412]: cannot find type `Error` in crate `stellar_xdr`
   --> /Users/willem/.cargo/git/checkouts/rs-soroban-env-230c62c6b1f3d3f1/0c0cae6/soroban-env-common/src/raw_val.rs:342:29
    |
342 |     fn from(_: stellar_xdr::Error) -> Self {
    |                             ^^^^^ not found in `stellar_xdr`
    |
help: consider importing one of these items
    |
1   + use core::error::Error;
    |
1   + use core::fmt::Error;
    |
1   + use crate::xdr::curr::Error;
    |
1   + use crate::xdr::next::Error;
    |
      and 11 other candidates
help: if you import `Error`, refer to it directly
    |
342 -     fn from(_: stellar_xdr::Error) -> Self {
342 +     fn from(_: Error) -> Self {
    |

error[E0412]: cannot find type `Error` in crate `stellar_xdr`
   --> /Users/willem/.cargo/git/checkouts/rs-soroban-env-230c62c6b1f3d3f1/0c0cae6/soroban-env-common/src/status.rs:165:31
    |
165 |     type Error = stellar_xdr::Error;
    |                               ^^^^^ not found in `stellar_xdr`
    |
help: consider importing one of these items
    |
1   + use core::error::Error;
    |
1   + use core::fmt::Error;
    |
1   + use crate::xdr::curr::Error;
    |
1   + use crate::xdr::next::Error;
    |
      and 11 other candidates
help: if you import `Error`, refer to it directly
    |
165 -     type Error = stellar_xdr::Error;
165 +     type Error = Error;
    |

error[E0412]: cannot find type `Error` in crate `stellar_xdr`
   --> /Users/willem/.cargo/git/checkouts/rs-soroban-env-230c62c6b1f3d3f1/0c0cae6/soroban-env-common/src/status.rs:195:31
    |
195 |     type Error = stellar_xdr::Error;
    |                               ^^^^^ not found in `stellar_xdr`
    |
help: consider importing one of these items
    |
1   + use core::error::Error;
    |
1   + use core::fmt::Error;
    |
1   + use crate::xdr::curr::Error;
    |
1   + use crate::xdr::next::Error;
    |
      and 11 other candidates
help: if you import `Error`, refer to it directly
    |
195 -     type Error = stellar_xdr::Error;
195 +     type Error = Error;
    |

error[E0412]: cannot find type `Error` in crate `stellar_xdr`
   --> /Users/willem/.cargo/git/checkouts/rs-soroban-env-230c62c6b1f3d3f1/0c0cae6/soroban-env-common/src/status.rs:272:24
    |
272 | impl From<stellar_xdr::Error> for Status {
    |                        ^^^^^ not found in `stellar_xdr`
    |
help: consider importing one of these items
    |
1   + use core::error::Error;
    |
1   + use core::fmt::Error;
    |
1   + use crate::xdr::curr::Error;
    |
1   + use crate::xdr::next::Error;
    |
      and 11 other candidates
help: if you import `Error`, refer to it directly
    |
272 - impl From<stellar_xdr::Error> for Status {
272 + impl From<Error> for Status {
    |

error[E0412]: cannot find type `Error` in crate `stellar_xdr`
   --> /Users/willem/.cargo/git/checkouts/rs-soroban-env-230c62c6b1f3d3f1/0c0cae6/soroban-env-common/src/status.rs:273:34
    |
273 |     fn from(_value: stellar_xdr::Error) -> Self {
    |                                  ^^^^^ not found in `stellar_xdr`
    |
help: consider importing one of these items
    |
1   + use core::error::Error;
    |
1   + use core::fmt::Error;
    |
1   + use crate::xdr::curr::Error;
    |
1   + use crate::xdr::next::Error;
    |
      and 11 other candidates
help: if you import `Error`, refer to it directly
    |
273 -     fn from(_value: stellar_xdr::Error) -> Self {
273 +     fn from(_value: Error) -> Self {
    |

error[E0531]: cannot find tuple struct or tuple variant `ScSymbol` in crate `crate::xdr`
   --> /Users/willem/.cargo/git/checkouts/rs-soroban-env-230c62c6b1f3d3f1/0c0cae6/soroban-env-common/src/symbol.rs:391:42
    |
391 |         if let ScVal::Symbol(crate::xdr::ScSymbol(vec)) = v {
    |                                          ^^^^^^^^ not found in `crate::xdr`
    |
help: consider importing one of these items
    |
1   + use crate::xdr::next::ScSymbol;
    |
1   + use crate::xdr::next::Type::ScSymbol;
    |
1   + use stellar_xdr::next::ScSymbol;
    |
1   + use stellar_xdr::next::Type::ScSymbol;
    |
help: if you import `ScSymbol`, refer to it directly
    |
391 -         if let ScVal::Symbol(crate::xdr::ScSymbol(vec)) = v {
391 +         if let ScVal::Symbol(ScSymbol(vec)) = v {
    |

Some errors have detailed explanations: E0412, E0425, E0432, E0433, E0531.
error: could not compile `soroban-env-common` (lib) due to 19 previous errors
```

**Comment by leighmcculloch** [comment]
Taking a look.

**Comment by leighmcculloch** [comment]
@willemneal The fix is this:
```diff
diff --git a/Cargo.toml b/Cargo.toml
index 0ef956a..feeb0a0 100644
--- a/Cargo.toml
+++ b/Cargo.toml
@@ -48,6 +48,7 @@ path = "cmd/soroban-cli"
 version = "0.0.16"
 git = "https://github.com/stellar/rs-stellar-xdr"
 rev = "53e1a9cf2335aff29305c72deb6f075e78915dad"
+default-features = false
 
 [workspace.dependencies]
 base64 = "0.13.0"
```

Workspace dependencies have this surprising feature that the workspace unifies with the crates, and the workspace is `default-features = true` by default.

**Comment by willemneal** [comment]
@leighmcculloch Thanks for your help! I also just fixed the issue that the new crates weren't being tested. But everything works locally for me now.

**Comment by willemneal** [comment]
@leighmcculloch Seems to be an issue on windows with the paths, which wasn't tested in the sdk.  Can we have the tests skipped on windows for now? I'll make an issue for it.

**Comment by leighmcculloch** [comment]
> @leighmcculloch Seems to be an issue on windows with the paths, which wasn't tested in the sdk. Can we have the tests skipped on windows for now? I'll make an issue for it.

Wdyt @tsachiherman? Not testing windows is a bit different to not shipping a precompiled binary for it. Assuming it's short lived is probably okay though and we can circle back on that soon.

**Comment by tsachiherman** [comment]
> > @leighmcculloch Seems to be an issue on windows with the paths, which wasn't tested in the sdk. Can we have the tests skipped on windows for now? I'll make an issue for it.
> 
> Wdyt @tsachiherman? Not testing windows is a bit different to not shipping a precompiled binary for it. Assuming it's short lived is probably okay though and we can circle back on that soon.

Sure. As long as we're not occurring too much tech debt, I see no reason to block that. Please open an issue to track that, and reference from within source code.

**Comment by willemneal** [comment]
@tsachiherman https://github.com/stellar/soroban-tools/issues/707

File: Makefile [hunk 1]
```diff
@@ -50,6 +50,9 @@ install: install_rust build-libpreflight
 build_rust: Cargo.lock
 	cargo build
 
+build-snapshot:
```
Comment (tsachiherman) [inline Makefile:53]: Add build-snapshop to the list of phony targets.

File: Makefile [hunk 2]
```diff
@@ -97,3 +100,4 @@ lint:
 
 # PHONY lists all the targets that aren't file names, so that make would skip the timestamp based check.
 .PHONY: publish clean fmt watch check e2e-test test build-test-wasms install build build-soroban-rpc build-libpreflight lint lint-changes
+
```
Comment (tsachiherman) [inline Makefile:103]: Drop redundant eol

File: cmd/crates/soroban-spec-typescript/fixtures/ts/.gitignore [hunk 3]
```diff
@@ -0,0 +1,2 @@
+node_modules/
+out/
```
Comment (tsachiherman) [inline cmd/crates/soroban-spec-typescript/fixtures/ts/.gitignore:2]: Add eol

**Review by tsachiherman — CHANGES_REQUESTED** [review]
Overall looks great, few minor requests.

**Review by tsachiherman — APPROVED** [review]

**Comment by leighmcculloch** [comment]
I'm removing soroban-spec-json from stellar/rs-soroban-sdk in https://github.com/stellar/rs-soroban-sdk/pull/992.

**Comment by leighmcculloch** [comment]
I merged the PR because I saw @tsachiherman approved it, and auto-merge was enabled, and it only got disabled because @willemneal pushed a tiny new line fix.

**Pull request merged**